Restructure docs, clarify extension points and configs, use short ids

jstriebel opened the pull request:
This PR

* restructures the headers to integrate other zarr pages,
* adds links to v2 and v1 specs,
* clarifies extension points, fixes #191:
  * future features will be added as new top-level keys
  * extension points have coherent configs
  * updated the stability policy accordingly
* changes all identifiers to be simple short forms, fixes #176,
* adds minor cleanups, e.g. removing resolved questions.

Preview of the built docs:
https://zarr-specs--204.org.readthedocs.build/en/204/v3/core/v3.0.html

jbms commented:
Note: complex numbers are now part of the main spec so we can remove that extension.

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -1112,40 +1065,26 @@ A Zarr group metadata object must contain the following mandatory key:
-    an appropriate error raised.
-
-``descenents_may_have_group_extensions``
+``descenents_must_be_read``
```
Note: My original name had a typo, "descenents" instead of "descendents".

However, I would actually propose that in the interest of getting v3 out sooner, we leave out this property for now so that we can consider it more carefully after getting out v3 but before adding any group extensions.  If we add this property later, and an implementation is unaware of it, it will fail if it encounters it, which is correct behavior, since if it doesn't know about this property then it also won't know about the group extensions.   There are some edge cases where technically the implementation might be able to read parts of the group successfully even without knowing about this property, but I don't think we have to worry about that.

jstriebel replied:
> However, I would actually propose that in the interest of getting v3 out sooner, we leave out this property for now so that we can consider it more carefully after getting out v3 but before adding any group extensions. If we add this property later, and an implementation is unaware of it, it will fail if it encounters it, which is correct behavior, since if it doesn't know about this property then it also won't know about the group extensions. There are some edge cases where technically the implementation might be able to read parts of the group successfully even without knowing about this property, but I don't think we have to worry about that.

Good idea! I added a separate issue for it: #207

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -1160,13 +1099,18 @@ Optional keys:
 
     {
-        "zarr_format": "https://purl.org/zarr/spec/v3#group",
+        "zarr_format": 3,
```
I think we need something to explicitly distinguish between arrays and groups, rather than having to do it based on the presence of data_type or other array properties.  Note that I had changed `zarr_format` to be different for arrays and groups ( "https://purl.org/zarr/spec/v3#array" vs "https://purl.org/zarr/spec/v3#group").

jstriebel replied:
I'd rather put that into a different key than the version, if there are no objections. How about `node_type`, which is either `group` or `array`?

jstriebel commented:
@jbms Thanks for the review! I applied your feedback, and also updated the sharding and codecs spec to the newest format. This is also necessary for the data-types, but I think those need to be partially re-written anyways, not only with the new configuration syntax. It would be great to have somebody authoring a ZEP for those. (cc @MSanKeys963, might be something to ask around for in the next community meeting?)

Also cc @joshmoore and @MSanKeys963, what do you think of the updated menu header? I could imagine to use the same headers with external links on multiple pages, to get the impression of a single coherent page. This doesn't need to be tackled in the specs first, but I thought this could be a good opportunity to play around with this idea, that's why I added it here.

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -844,18 +822,21 @@ encoded for storage.
 
-    Must be the string "https://purl.org/zarr/spec/v3#array".
+    An integer defining the version of the storage specification to which the
+    array store adheres, muste be ``3`` here.
```
type on "muste"

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -844,18 +822,21 @@ encoded for storage.
+``node_type``
+^^^^^^^^^^^^^^^
+
+    A string defining the type of hierarchy node element, muste be ``array``
```
typo on "muste"

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -1112,43 +1074,21 @@ A Zarr group metadata object must contain the following mandatory key:
+``node_type``
+^^^^^^^^^^^^^^^
+
+    A string defining the type of hierarchy node element, muste be ``group``
```
typo on "muste"

jbms reviewed (dismissed):
Aside from the typos, looks good to me.  We may want to iterate more, but easier if this is submitted first, I think.
